rasa_nlu_examples/classifiers/sparse_sklearn_intent_classifier.py

```diff
@@ -0,0 +1,275 @@
+import typing
+import warnings
+from typing import Any, Dict, List, Optional, Text, Tuple, Type
+
+import numpy as np
+import scipy.sparse
+
+import rasa.shared.utils.io
+import rasa.utils.io as io_utils
+from rasa.shared.constants import DOCS_URL_TRAINING_DATA_NLU
+from rasa.nlu.classifiers import LABEL_RANKING_LENGTH
+from rasa.nlu.featurizers.featurizer import SparseFeaturizer
+from rasa.nlu.components import Component
+from rasa.nlu.classifiers.classifier import IntentClassifier
+from rasa.nlu.config import RasaNLUModelConfig
+from rasa.shared.nlu.constants import TEXT
+from rasa.nlu.model import Metadata
+from rasa.shared.nlu.training_data.training_data import TrainingData
+from rasa.shared.nlu.training_data.message import Message
+from rasa.shared.nlu.constants import (
+    INTENT,
+    INTENT_NAME_KEY,
+    PREDICTED_CONFIDENCE_KEY,
+)
+
+logger = logging.getLogger(__name__)
+
+if typing.TYPE_CHECKING:
+    import sklearn
+
+
+class SparseSklearnIntentClassifier(IntentClassifier):
+    r"""Base class for intent classifiers using the sklearn framework with sparse features.
+    Note that all sparse features will be used, i.e. there is no filtering for specific
+    featurizers yet."""
+
+    defaults = {}
+
+    @classmethod
+    def required_packages(cls) -> List[Text]:
+        return ["sklearn"]
+
+    @classmethod
+    def required_components(cls) -> List[Type[Component]]:
+        return [SparseFeaturizer]
+
+    def __init__(
+        self,
+        component_config: Optional[Dict[Text, Any]] = None,
+        clf: Optional["sklearn.base.ClassifierMixin"] = None,
+        le: Optional["sklearn.preprocessing.LabelEncoder"] = None,
+    ) -> None:
+        """Construct a new naive Bayes intent classifier using the sklearn framework."""
+        super().__init__(component_config)
+
+        if (not clf) + (not le) == 1:
+            raise ValueError(
+                "Expected classifier and label encoder instance. "
+                "Specifying only one of them leads to an incomplete model definition."
+            )
+        self.le = le
+        self.clf = clf
+
+    @abstractmethod
+    def create_sklearn_classifier(
+        self, **kwargs: Any
+    ) -> "sklearn.base.ClassifierMixin":
+        r"""Lazily imports the required sklearn classifier class and creates and
+        instantiates the sklearn classifier using all the given keyword arguments.
+
+        :param **kwargs: see defaults dictionary
+        """
+        pass
+
+    def train(
+        self,
+        training_data: TrainingData,
+        config: Optional[RasaNLUModelConfig] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Trains the intent classifier from scratch on a data set. A any classifier
+        components passed to the constructor will be forgotten, i.e.
+        you cannot use it for finetuning yet.
+
+        Reminder: Enabling finetuning here would be possible for sklearn
+        models that implement partial_fit or have a warm_start option.
```

Comment:
Merely as a mention in case you're interested. In general, scikit-learn doesn't do much with the `.partial_fit` functionality. Their `Pipeline` class doesn't allow for it. If you're interested in these features though I do maintain [tokenwiser](https://github.com/koaning/tokenwiser) in an attempt to add these features to the sklearn ecosystem.